Make Diagnostic a trait

```
error[E0038]: the trait `diagnostic::Diagnostic` cannot be made into an object
  --> src/append/mod.rs:52:28
   |
52 |         diagnostics: &[Box<dyn Diagnostic>],
   |                            ^^^^^^^^^^^^^^ `diagnostic::Diagnostic` cannot be made into an object
   |
note: for a trait to be "dyn-compatible" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
  --> src/diagnostic/mod.rs:42:8
   |
40 | pub trait Diagnostic: fmt::Debug + Send + Sync + 'static {
   |           ---------- this trait cannot be made into an object...
41 |     /// Visits the diagnostic key-value pairs.
42 |     fn visit<V: Visitor>(&self, visitor: &mut V);
   |        ^^^^^ ...because method `visit` has generic type parameters
   = help: consider moving `visit` to another trait
   = help: the following types implement the trait, consider defining an enum where each variant holds one of these types, implementing `diagnostic::Diagnostic` for this new enum and using it instead:
             diagnostic::fastrace::FastraceDiagnostic
             diagnostic::static_global::StaticDiagnostic
             diagnostic::thread_local::ThreadLocalDiagnostic
   = note: `diagnostic::Diagnostic` can be implemented in other crates; if you want to support your users passing their own types here, you can't refer to a specific type
```

Doesn't seem to work. Perhaps add a `FnDiagnostic` for accepting custom logics.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -50,7 +50,7 @@ All notable changes to this project will be documented in this file.
 ### Breaking changes
 
 * Bump MSRV to 1.80 for upgrading `colored` to 3.0.
-* Layout and Appender now accept a new argument `diagnostics: &[Diagnostic]` to retrieve the mapped diagnostics context. Users can use `logforth::diagnostic::Visitor` to visit the diagnostics context.
+* Layout and Appender now accept a new argument `diagnostics: &[Box<dyn Diagnostic>]` to retrieve the mapped diagnostics context. Users can use `logforth::diagnostic::Visitor` to visit the diagnostics context.
 
 ### New features
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,10 @@ All notable changes to this project will be documented in this file.
 
 ## Unreleased
 
+### Breaking changes
+
+* `Diagnosic` is now a trait. `Visitor`'s method signature is simplified.
+
 ### New features
 
 * Add `BlockingRollingFile` and `BlockingSyslog` appenders. They would lock the inner appender and log synchronously. In some lower rate logging scenarios, this would be more efficient than the non-blocking version.
@@ -50,7 +54,7 @@ All notable changes to this project will be documented in this file.
 ### Breaking changes
 
 * Bump MSRV to 1.80 for upgrading `colored` to 3.0.
-* Layout and Appender now accept a new argument `diagnostics: &[Box<dyn Diagnostic>]` to retrieve the mapped diagnostics context. Users can use `logforth::diagnostic::Visitor` to visit the diagnostics context.
+* Layout and Appender now accept a new argument `diagnostics: &[Diagnostic]` to retrieve the mapped diagnostics context. Users can use `logforth::diagnostic::Visitor` to visit the diagnostics context.
 
 ### New features
 

--- a/examples/custom_layout_filter.rs
+++ b/examples/custom_layout_filter.rs
@@ -37,7 +37,7 @@ impl Filter for CustomFilter {
 struct CustomLayout;
 
 impl Layout for CustomLayout {
-    fn format(&self, record: &Record, _: &[Diagnostic]) -> anyhow::Result<Vec<u8>> {
+    fn format(&self, record: &Record, _: &[Box<dyn Diagnostic>]) -> anyhow::Result<Vec<u8>> {
         Ok(format!("[Alert] {}", record.args()).into_bytes())
     }
 }

--- a/src/append/fastrace.rs
+++ b/src/append/fastrace.rs
@@ -38,7 +38,7 @@ pub struct FastraceEvent {
 }
 
 impl Append for FastraceEvent {
-    fn append(&self, record: &Record, diagnostics: &[Diagnostic]) -> anyhow::Result<()> {
+    fn append(&self, record: &Record, diagnostics: &[Box<dyn Diagnostic>]) -> anyhow::Result<()> {
         let message = format!("{}", record.args());
 
         let mut collector = KvCollector { kv: Vec::new() };
@@ -87,13 +87,9 @@ impl<'kvs> log::kv::VisitSource<'kvs> for KvCollector {
 }
 
 impl Visitor for KvCollector {
-    fn visit<'k, 'v, K, V>(&mut self, key: K, value: V)
-    where
-        K: Into<Cow<'k, str>>,
-        V: Into<Cow<'v, str>>,
-    {
-        let key = key.into().into_owned();
-        let value = value.into().into_owned();
+    fn visit(&mut self, key: Cow<str>, value: Cow<str>) {
+        let key = key.into_owned();
+        let value = value.into_owned();
         self.kv.push((key, value));
     }
 }

--- a/src/append/journald/mod.rs
+++ b/src/append/journald/mod.rs
@@ -247,13 +247,7 @@ impl<'kvs> log::kv::VisitSource<'kvs> for WriteKeyValues<'_> {
 }
 
 impl Visitor for WriteKeyValues<'_> {
-    fn visit<'k, 'v, K, V>(&mut self, key: K, value: V)
-    where
-        K: Into<Cow<'k, str>>,
-        V: Into<Cow<'v, str>>,
-    {
-        let key = key.into();
-        let value = value.into();
+    fn visit(&mut self, key: Cow<str>, value: Cow<str>) {
         let key = key.as_ref();
         let value = value.as_bytes();
         field::put_field_length_encoded(self.0, field::FieldName::WriteEscaped(key), value);
@@ -263,7 +257,7 @@ impl Visitor for WriteKeyValues<'_> {
 impl Append for Journald {
     /// Extract all fields (standard and custom) from `record`, append all `extra_fields` given
     /// to this appender, and send the result to journald.
-    fn append(&self, record: &Record, diagnostics: &[Diagnostic]) -> anyhow::Result<()> {
+    fn append(&self, record: &Record, diagnostics: &[Box<dyn Diagnostic>]) -> anyhow::Result<()> {
         use field::*;
 
         let mut buffer = vec![];

--- a/src/append/mod.rs
+++ b/src/append/mod.rs
@@ -50,7 +50,11 @@ pub use self::syslog::Syslog;
 /// A trait representing an appender that can process log records.
 pub trait Append: fmt::Debug + Send + Sync + 'static {
     /// Dispatches a log record to the append target.
-    fn append(&self, record: &log::Record, diagnostics: &[Diagnostic]) -> anyhow::Result<()>;
+    fn append(
+        &self,
+        record: &log::Record,
+        diagnostics: &[Box<dyn Diagnostic>],
+    ) -> anyhow::Result<()>;
 
     /// Flushes any buffered records.
     fn flush(&self) {}

--- a/src/append/opentelemetry.rs
+++ b/src/append/opentelemetry.rs
@@ -235,7 +235,7 @@ pub struct OpentelemetryLog {
 }
 
 impl Append for OpentelemetryLog {
-    fn append(&self, record: &Record, diagnostics: &[Diagnostic]) -> anyhow::Result<()> {
+    fn append(&self, record: &Record, diagnostics: &[Box<dyn Diagnostic>]) -> anyhow::Result<()> {
         let mut log_record = self.logger.create_log_record();
         log_record.set_observed_timestamp(SystemTime::now());
         log_record.set_severity_number(log_level_to_otel_severity(record.level()));
@@ -303,13 +303,9 @@ impl<'kvs> log::kv::VisitSource<'kvs> for KvExtractor<'_> {
 }
 
 impl Visitor for KvExtractor<'_> {
-    fn visit<'k, 'v, K, V>(&mut self, key: K, value: V)
-    where
-        K: Into<Cow<'k, str>>,
-        V: Into<Cow<'v, str>>,
-    {
-        let key = key.into().into_owned();
-        let value = value.into().into_owned();
+    fn visit(&mut self, key: Cow<str>, value: Cow<str>) {
+        let key = key.into_owned();
+        let value = value.into_owned();
         self.record.add_attribute(key, value);
     }
 }

--- a/src/append/rolling_file/append.rs
+++ b/src/append/rolling_file/append.rs
@@ -50,7 +50,7 @@ impl BlockingRollingFile {
 }
 
 impl Append for BlockingRollingFile {
-    fn append(&self, record: &Record, diagnostics: &[Diagnostic]) -> anyhow::Result<()> {
+    fn append(&self, record: &Record, diagnostics: &[Box<dyn Diagnostic>]) -> anyhow::Result<()> {
         let mut bytes = self.layout.format(record, diagnostics)?;
         bytes.push(b'\n');
         let mut writer = self.writer.lock().unwrap_or_else(|e| e.into_inner());
@@ -92,7 +92,7 @@ impl RollingFile {
 }
 
 impl Append for RollingFile {
-    fn append(&self, record: &Record, diagnostics: &[Diagnostic]) -> anyhow::Result<()> {
+    fn append(&self, record: &Record, diagnostics: &[Box<dyn Diagnostic>]) -> anyhow::Result<()> {
         let mut bytes = self.layout.format(record, diagnostics)?;
         bytes.push(b'\n');
         self.writer.send(bytes)?;

--- a/src/append/stdio.rs
+++ b/src/append/stdio.rs
@@ -61,7 +61,7 @@ impl Stdout {
 }
 
 impl Append for Stdout {
-    fn append(&self, record: &Record, diagnostics: &[Diagnostic]) -> anyhow::Result<()> {
+    fn append(&self, record: &Record, diagnostics: &[Box<dyn Diagnostic>]) -> anyhow::Result<()> {
         let mut bytes = self.layout.format(record, diagnostics)?;
         bytes.push(b'\n');
         std::io::stdout().write_all(&bytes)?;
@@ -113,7 +113,7 @@ impl Stderr {
 }
 
 impl Append for Stderr {
-    fn append(&self, record: &Record, diagnostics: &[Diagnostic]) -> anyhow::Result<()> {
+    fn append(&self, record: &Record, diagnostics: &[Box<dyn Diagnostic>]) -> anyhow::Result<()> {
         let mut bytes = self.layout.format(record, diagnostics)?;
         bytes.push(b'\n');
         std::io::stderr().write_all(&bytes)?;

--- a/src/append/syslog.rs
+++ b/src/append/syslog.rs
@@ -106,7 +106,7 @@ impl Syslog {
 }
 
 impl Append for Syslog {
-    fn append(&self, record: &Record, diagnostics: &[Diagnostic]) -> anyhow::Result<()> {
+    fn append(&self, record: &Record, diagnostics: &[Box<dyn Diagnostic>]) -> anyhow::Result<()> {
         let message = self.formatter.format_message(record, diagnostics)?;
         self.writer.send(message)?;
         Ok(())
@@ -155,7 +155,7 @@ impl BlockingSyslog {
 }
 
 impl Append for BlockingSyslog {
-    fn append(&self, record: &Record, diagnostics: &[Diagnostic]) -> anyhow::Result<()> {
+    fn append(&self, record: &Record, diagnostics: &[Box<dyn Diagnostic>]) -> anyhow::Result<()> {
         let message = self.formatter.format_message(record, diagnostics)?;
         let mut writer = self.writer.lock().unwrap_or_else(|e| e.into_inner());
         writer.write_all(message.as_slice())?;
@@ -191,7 +191,7 @@ impl SyslogFormatter {
     fn format_message(
         &self,
         record: &Record,
-        diagnostics: &[Diagnostic],
+        diagnostics: &[Box<dyn Diagnostic>],
     ) -> anyhow::Result<Vec<u8>> {
         let severity = log_level_to_otel_severity(record.level());
 

--- a/src/diagnostic/fastrace.rs
+++ b/src/diagnostic/fastrace.rs
@@ -39,17 +39,11 @@ use crate::Diagnostic;
 #[non_exhaustive]
 pub struct FastraceDiagnostic {}
 
-impl FastraceDiagnostic {
-    pub fn visit<V: Visitor>(&self, visitor: &mut V) {
+impl Diagnostic for FastraceDiagnostic {
+    fn visit(&self, visitor: &mut dyn Visitor) {
         if let Some(span) = fastrace::collector::SpanContext::current_local_parent() {
             let trace_id = format!("{:016x}", span.trace_id.0);
-            visitor.visit("trace_id", trace_id);
+            visitor.visit("trace_id".into(), trace_id.into());
         }
-    }
-}
-
-impl From<FastraceDiagnostic> for Diagnostic {
-    fn from(diagnostic: FastraceDiagnostic) -> Self {
-        Diagnostic::Fastrace(diagnostic)
     }
 }

--- a/src/diagnostic/mod.rs
+++ b/src/diagnostic/mod.rs
@@ -16,6 +16,7 @@
 //! request.
 
 use std::borrow::Cow;
+use std::fmt;
 
 #[cfg(feature = "fastrace")]
 pub use self::fastrace::FastraceDiagnostic;
@@ -30,29 +31,16 @@ mod thread_local;
 /// A visitor to walk through diagnostic key-value pairs.
 pub trait Visitor {
     /// Visits a key-value pair.
-    fn visit<'k, 'v, K, V>(&mut self, key: K, value: V)
-    where
-        K: Into<Cow<'k, str>>,
-        V: Into<Cow<'v, str>>;
+    fn visit(&mut self, key: Cow<str>, value: Cow<str>);
 }
 
-/// Represent a Mapped Diagnostic Context (MDC) that provides diagnostic key-values.
-#[derive(Debug)]
-pub enum Diagnostic {
-    #[cfg(feature = "fastrace")]
-    Fastrace(FastraceDiagnostic),
-    Static(StaticDiagnostic),
-    ThreadLocal(ThreadLocalDiagnostic),
+/// A trait representing a Mapped Diagnostic Context (MDC) that provides diagnostic key-values.
+pub trait Diagnostic: fmt::Debug + Send + Sync + 'static {
+    fn visit(&self, visitor: &mut dyn Visitor);
 }
 
-impl Diagnostic {
-    /// Visits the diagnostic key-value pairs.
-    pub fn visit<V: Visitor>(&self, visitor: &mut V) {
-        match self {
-            #[cfg(feature = "fastrace")]
-            Diagnostic::Fastrace(diagnostic) => diagnostic.visit(visitor),
-            Diagnostic::Static(diagnostic) => diagnostic.visit(visitor),
-            Diagnostic::ThreadLocal(diagnostic) => diagnostic.visit(visitor),
-        }
+impl<T: Diagnostic> From<T> for Box<dyn Diagnostic> {
+    fn from(value: T) -> Self {
+        Box::new(value)
     }
 }

--- a/src/diagnostic/static_global.rs
+++ b/src/diagnostic/static_global.rs
@@ -49,16 +49,12 @@ impl StaticDiagnostic {
     pub fn remove(&mut self, key: &str) {
         self.kvs.remove(key);
     }
-
-    pub fn visit<V: Visitor>(&self, visitor: &mut V) {
-        for (key, value) in self.kvs.iter() {
-            visitor.visit(key, value);
-        }
-    }
 }
 
-impl From<StaticDiagnostic> for Diagnostic {
-    fn from(diagnostic: StaticDiagnostic) -> Self {
-        Diagnostic::Static(diagnostic)
+impl Diagnostic for StaticDiagnostic {
+    fn visit(&self, visitor: &mut dyn Visitor) {
+        for (key, value) in self.kvs.iter() {
+            visitor.visit(key.into(), value.into());
+        }
     }
 }

--- a/src/diagnostic/thread_local.rs
+++ b/src/diagnostic/thread_local.rs
@@ -51,19 +51,15 @@ impl ThreadLocalDiagnostic {
             map.borrow_mut().remove(key);
         });
     }
+}
 
-    pub fn visit<V: Visitor>(&self, visitor: &mut V) {
+impl Diagnostic for ThreadLocalDiagnostic {
+    fn visit(&self, visitor: &mut dyn Visitor) {
         CONTEXT.with(|map| {
             let map = map.borrow();
             for (key, value) in map.iter() {
-                visitor.visit(key, value);
+                visitor.visit(key.into(), value.into());
             }
         })
-    }
-}
-
-impl From<ThreadLocalDiagnostic> for Diagnostic {
-    fn from(diagnostic: ThreadLocalDiagnostic) -> Self {
-        Diagnostic::ThreadLocal(diagnostic)
     }
 }

--- a/src/layout/json.rs
+++ b/src/layout/json.rs
@@ -86,13 +86,9 @@ impl<'kvs> log::kv::VisitSource<'kvs> for KvCollector<'_> {
 }
 
 impl Visitor for KvCollector<'_> {
-    fn visit<'k, 'v, K, V>(&mut self, key: K, value: V)
-    where
-        K: Into<Cow<'k, str>>,
-        V: Into<Cow<'v, str>>,
-    {
-        let key = key.into().into_owned();
-        let value = value.into().into_owned();
+    fn visit(&mut self, key: Cow<str>, value: Cow<str>) {
+        let key = key.into_owned();
+        let value = value.into_owned();
         self.kvs.insert(key, value.into());
     }
 }
@@ -125,7 +121,11 @@ where
 }
 
 impl Layout for JsonLayout {
-    fn format(&self, record: &Record, diagnostics: &[Diagnostic]) -> anyhow::Result<Vec<u8>> {
+    fn format(
+        &self,
+        record: &Record,
+        diagnostics: &[Box<dyn Diagnostic>],
+    ) -> anyhow::Result<Vec<u8>> {
         let mut kvs = Map::new();
         let mut visitor = KvCollector { kvs: &mut kvs };
         record.key_values().visit(&mut visitor)?;

--- a/src/layout/mod.rs
+++ b/src/layout/mod.rs
@@ -29,7 +29,11 @@ pub use text::TextLayout;
 /// A layout for formatting log records.
 pub trait Layout: fmt::Debug + Send + Sync + 'static {
     /// Formats a log record with optional diagnostics.
-    fn format(&self, record: &log::Record, diagnostics: &[Diagnostic]) -> anyhow::Result<Vec<u8>>;
+    fn format(
+        &self,
+        record: &log::Record,
+        diagnostics: &[Box<dyn Diagnostic>],
+    ) -> anyhow::Result<Vec<u8>>;
 }
 
 impl<T: Layout> From<T> for Box<dyn Layout> {

--- a/src/layout/text.rs
+++ b/src/layout/text.rs
@@ -154,7 +154,11 @@ impl TextLayout {
 }
 
 impl Layout for TextLayout {
-    fn format(&self, record: &Record, diagnostics: &[Diagnostic]) -> anyhow::Result<Vec<u8>> {
+    fn format(
+        &self,
+        record: &Record,
+        diagnostics: &[Box<dyn Diagnostic>],
+    ) -> anyhow::Result<Vec<u8>> {
         let time = match self.tz.clone() {
             Some(tz) => Timestamp::now().to_zoned(tz),
             None => Zoned::now(),
@@ -204,19 +208,9 @@ impl<'kvs> log::kv::VisitSource<'kvs> for KvWriter {
 }
 
 impl Visitor for KvWriter {
-    fn visit<'k, 'v, K, V>(&mut self, key: K, value: V)
-    where
-        K: Into<Cow<'k, str>>,
-        V: Into<Cow<'v, str>>,
-    {
+    fn visit(&mut self, key: Cow<str>, value: Cow<str>) {
         // SAFETY: no more than an allocate-less version
         //  self.text.push_str(&format!(" {key}={value}"))
-        write!(
-            &mut self.text,
-            " {key}={value}",
-            key = key.into(),
-            value = value.into()
-        )
-        .unwrap();
+        write!(&mut self.text, " {key}={value}").unwrap();
     }
 }

--- a/src/logger/builder.rs
+++ b/src/logger/builder.rs
@@ -194,7 +194,7 @@ impl Builder {
 #[derive(Debug)]
 pub struct DispatchBuilder<const APPEND: bool> {
     filters: Vec<Box<dyn Filter>>,
-    diagnostics: Vec<Diagnostic>,
+    diagnostics: Vec<Box<dyn Diagnostic>>,
     appends: Vec<Box<dyn Append>>,
 }
 
@@ -242,7 +242,7 @@ impl DispatchBuilder<false> {
     ///     })
     ///     .apply();
     /// ```
-    pub fn diagnostic(mut self, diagnostic: impl Into<Diagnostic>) -> Self {
+    pub fn diagnostic(mut self, diagnostic: impl Into<Box<dyn Diagnostic>>) -> Self {
         self.diagnostics.push(diagnostic.into());
         self
     }

--- a/src/logger/log_impl.rs
+++ b/src/logger/log_impl.rs
@@ -69,14 +69,14 @@ impl log::Log for Logger {
 #[derive(Debug)]
 pub(super) struct Dispatch {
     filters: Vec<Box<dyn Filter>>,
-    diagnostics: Vec<Diagnostic>,
+    diagnostics: Vec<Box<dyn Diagnostic>>,
     appends: Vec<Box<dyn Append>>,
 }
 
 impl Dispatch {
     pub(super) fn new(
         filters: Vec<Box<dyn Filter>>,
-        diagnostics: Vec<Diagnostic>,
+        diagnostics: Vec<Box<dyn Diagnostic>>,
         appends: Vec<Box<dyn Append>>,
     ) -> Self {
         debug_assert!(

--- a/tests/recursive_logging.rs
+++ b/tests/recursive_logging.rs
@@ -26,7 +26,7 @@ use logforth::Layout;
 struct CustomLayout(&'static str);
 
 impl Layout for CustomLayout {
-    fn format(&self, record: &Record, _: &[Diagnostic]) -> anyhow::Result<Vec<u8>> {
+    fn format(&self, record: &Record, _: &[Box<dyn Diagnostic>]) -> anyhow::Result<Vec<u8>> {
         Ok(format!("{} [{}] {}", self.0, record.level(), record.args()).into_bytes())
     }
 }

